File: runtime/org.openmole.runtime.daemon/src/main/java/org/openmole/runtime/daemon/Daemon.java
/*
 *  Copyright (C) 2010 Romain Reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.runtime.daemon;

import org.openmole.core.batch.authentication.Authentication;
import java.io.File;
import java.util.logging.Level;

import java.util.logging.Logger;
import org.apache.commons.cli.BasicParser;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;

import org.eclipse.equinox.app.IApplication;
import org.eclipse.equinox.app.IApplicationContext;
import org.openmole.core.batch.file.URIFile;
import org.openmole.core.batch.authentication.JSAGASessionService;
import org.openmole.misc.pluginmanager.PluginManager;
import org.openmole.misc.workspace.Workspace;
import org.openmole.core.serializer.SerializerService;

public class Daemon implements IApplication {
    
    @Override
    public Object start(IApplicationContext context) throws Exception {
        try {
            System.setProperty("org.openmole.misc.workspace.noUniqueResource", "true");
            String args[] = (String[]) context.getArguments().get(IApplicationContext.APPLICATION_ARGS);
            
            Options options = new Options();
            
            options.addOption("h", true, "user@hostname:port");
            options.addOption("p", true, "password");
            options.addOption("w", true, "number of workers");
            options.addOption("d", false, "debug mode");
            options.addOption("c", true, "cache size in Mo (default is 2000)");
            
            CommandLineParser parser = new BasicParser();
            CommandLine cmdLine;
            
            try {
                cmdLine = parser.parse(options, args);
            } catch (ParseException e) {
                Logger.getLogger(Daemon.class.getName()).severe("Error while parsing command line arguments");
                new HelpFormatter().printHelp(" ", options);
                return IApplication.EXIT_OK;
            }
            
            String userHostnamePort = cmdLine.getOptionValue("h");
            String password = cmdLine.getOptionValue("p");
            boolean debug = cmdLine.hasOption("d");
            
            int workers = 1;
            
            if(cmdLine.hasOption("w")) workers = new Integer(cmdLine.getOptionValue("w"));
            
            if (userHostnamePort == null || password == null) {
                Logger.getLogger(Daemon.class.getName()).severe("Error while parsing command line arguments");
                new HelpFormatter().printHelp(" ", options);
                return IApplication.EXIT_OK;                
            }
            
            long cacheSize = 2000;
            if(cmdLine.hasOption("c")) cacheSize = new Long(cmdLine.getOptionValue("c"));
            
            new JobLauncher(cacheSize * 1024 * 1024, debug).launch(userHostnamePort, password, workers);
            
        } catch (Throwable t) {
            Logger.getLogger(Daemon.class.getName()).log(Level.SEVERE, "Error durring runtime execution", t);
        }
        return IApplication.EXIT_OK;
    }
    
    @Override
    public void stop() {
    }
}

File: runtime/org.openmole.runtime.daemon/src/main/scala/org/openmole/runtime/daemon/Daemon.scala
/*
 * Copyright (C) 2012 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.runtime.daemon

import org.eclipse.equinox.app._
import org.openmole.misc.tools.service.Logger
import scopt.immutable._

class Daemon extends IApplication with Logger {
  override def start(context: IApplicationContext) = {
    try {
      val args = context.getArguments.get(IApplicationContext.APPLICATION_ARGS).asInstanceOf[Array[String]]

      case class Config(
        host: Option[String] = None,
        password: Option[String] = None,
        workers: Int = 1,
        cacheSize: Int = 2000)

      val parser = new OptionParser[Config]("openmole", "0.x") {
        def options = Seq(
          opt("h", "host", "user@hostname:port") {
            (v: String, c: Config) ⇒ c.copy(host = Some(v))
          },
          opt("p", "password", "password") {
            (v: String, c: Config) ⇒ c.copy(password = Some(v))
          },
          opt("w", "workers", "Number of workers, default is 1") {
            (v: String, c: Config) ⇒ c.copy(workers = v.toInt)
          },
          opt("c", "cache", "Chache size, default is 2000") {
            (v: String, c: Config) ⇒ c.copy(cacheSize = v.toInt)
          })
      }

      val debug = args.contains("-d")
      val filtredArgs = args.filterNot((_: String) == "-d")
      parser.parse(filtredArgs, Config()) foreach { config ⇒
        new JobLauncher(config.cacheSize * 1024 * 1024, debug).launch(
          config.host.getOrElse(throw new RuntimeException("Host undifined")),
          config.password.getOrElse(throw new RuntimeException("Password undifined")),
          config.workers)
      }
    } catch {
      case t: Throwable ⇒ logger.log(SEVERE, "Error durring runtime execution", t)
    }
    IApplication.EXIT_OK
  }

  override def stop = {}

}
